test_util/recordio.py

```diff
@@ -0,0 +1,150 @@
+        """
+
+        s = self.serialize(message)
+
+        if not isinstance(s, bytes):
+            raise Exception("Calling 'serialize(message)' must return a 'bytes' object")
+
+        return bytes(str(len(s)) + "\n", "UTF-8") + s
+
+
+class Decoder():
+    """Decode a 'RecordIO' message back to an arbitrary message type.
+
+       This class encapsulates the process of decoding a message
+       previously encoded with 'RecordIO' back to an arbitrary
+       message type. Its constructor takes a deserialization
+       function of the form 'deserialize(data)'. This
+       deserialization function is responsible for knowing how to
+       take a fully constructed 'RecordIO' message containing a
+       'UTF-8' encoded byte array and deserialize it back into the
+       original message type.
+
+       The 'decode(data)' message takes a 'UTF-8' encoded byte array
+       as input and buffers it across subsequent calls to
+       construct a set of fully constructed 'RecordIO' messages that
+       are decoded and returned in a list.
+
+       :param deserialize: a function to deserialize from 'RecordIO'
+                           messages built up by subsequent calls
+                           to 'decode(data)'
+       :type deserialize: function
+    """
+
+    HEADER = 0
+    RECORD = 1
+    FAILED = 2
+
+    def __init__(self, deserialize):
+        self.deserialize = deserialize
+        self.state = self.HEADER
+        self.buffer = bytes("", "UTF-8")
+        self.length = 0
+
+    def decode(self, data):
+        """Decode a 'RecordIO' formatted message to its original type.
+
+        :param data: an array of 'UTF-8' encoded bytes that make up a
+                      partial 'RecordIO' message. Subsequent calls to this
+                      function maintain state to build up a full 'RecordIO'
+                      message and decode it
+        :type data: bytes
+        :returns: a list of deserialized messages
+        :rtype: list
+        """
+
+        if not isinstance(data, bytes):
+            raise Exception("Parameter 'data' must of of type 'bytes'")
+
+        if self.state == self.FAILED:
+            raise Exception("Decoder is in a FAILED state")
+
+        records = []
+
+        for c in data:
+            if self.state == self.HEADER:
+                if c != ord('\n'):
+                    self.buffer += bytes([c])
+                    continue
+
+                try:
+                    self.length = int(self.buffer.decode("UTF-8"))
+                except Exception as exception:
+                    self.state = self.FAILED
+                    raise Exception("Failed to decode length '{buffer}': {error}"
+                                    .format(buffer=self.buffer, error=exception))
+
+                self.buffer = bytes("", "UTF-8")
+                self.state = self.RECORD
+
+                # Note that for 0 length records, we immediately decode.
+                if self.length <= 0:
```

Comment:
Let's assert length is nonnegative after we decode it, to fail earlier on bad encodings and simplify length comparisons.